lib/rate_map.c

```diff
@@ -203,3 +210,95 @@ rate_map_shift_by_mass(rate_map_t *self, double pos, double mass)
     double result_mass = rate_map_position_to_mass(self, pos) + mass;
     return rate_map_mass_to_position(self, result_mass);
 }
+
+#ifdef _Static_assert
+_Static_assert(FLT_RADIX == 2, "Base 2 floating point types required");
+#endif
+
+static int
+fast_search_lookup_init_lookups(
+    fast_search_lookup_t *self, const double *values, size_t n_values)
+{
+    int ret = 0;
+    const double *lower, *strict_upper;
+    uint32_t i;
+    double query = -INFINITY;
+
+    lower = values;
+    for (i = 0; i < self->num_lookups; i++) {
+        while (*lower <= query) {
+            lower++;
+        }
+        self->lookups[i].start = lower;
+        strict_upper = lower;
+        query = ldexp(i, -(self->power_shift));
+        assert(strict_upper < values + n_values);
+        while (strict_upper < values + n_values && *strict_upper <= query) {
+            strict_upper++;
+        }
+        const int64_t num = strict_upper - lower;
+        if (num > INT32_MAX) {
+            ret = MSP_ERR_BAD_PARAM_VALUE; // TODO: consider new err code
+            goto out;
+        }
+        self->lookups[i].num = (size_t) num;
+    }
+out:
+    return ret;
+}
+
+static inline long
+ceil_to_long(double x)
+{
+    return lrint(ceil(x)); // TODO: test use of fast non-portable equivalent
+}
+
+int
+fast_search_lookup_alloc(
+    fast_search_lookup_t *self, const double *values, size_t n_values)
+{
+    int ret = 0;
+    const double max = values[n_values - 1];
+    const int power = ilogb((double) n_values) + 1;
+
+    if (*values != 0.0 || !isfinite(max) || max < 0) {
+        ret = MSP_ERR_BAD_PARAM_VALUE;
+        goto out;
+    }
+    self->power_shift = power - (ilogb(max) + 1);
+    const long max_index = ceil_to_long(ldexp(max, self->power_shift));
+    self->num_lookups = 1 + (size_t) max_index;
+    self->lookups = malloc(self->num_lookups * sizeof(*(self->lookups)));
+    ret = fast_search_lookup_init_lookups(self, values, n_values);
+out:
+    if (ret != 0) {
```

Comment:
We don't need to do this conditional free here. The API contract is that once ``x_alloc(&obj)`` has been called the client *must* call `x_free(&obj)``. It is always safe to call x_free after x_init has been called (and hence, x_init must take care to zero out any pointers at the start.